src/app/content/utils/seoUtils.ts

```diff
@@ -9,20 +12,182 @@ import {
+    const name = getArchiveTreeSectionTitle(node);
+    return includeTitle ? `Ch. ${number}. ${name}` : `Ch. ${number} `;
   }
 
   return archiveTreeSectionIsBook(node.parent)
     ? getArchiveTreeSectionTitle(node) + ' '
-    : getParentPrefix(node.parent);
+    : getParentPrefix(node.parent, includeTitle);
+
+};
+
+const hideMath = (node: Element) => {
+  if (!node) {
+    return '';
+  }
+
+  node.querySelectorAll('.os-math-in-para').forEach((el: Element) => {
+    el.outerHTML = '...';
+  });
+  return node.textContent;
+};
+
+export const getTextContent = (str: string) => {
+  const parsed = domParser.parseFromString(str, 'text/html');
+  const text = parsed.body.textContent;
+  return text;
+};
+
+const removeIntroContent = (node: HTMLElement) => {
+  if (!node) {
+    return null;
+  }
+
+  const introContentList = node.querySelectorAll(
+    '[data-type="abstract"], .learning-objectives, .chapter-objectives, .be-prepared, .os-teacher'
+    ) || [];
+
+  for (let i = 0; i <= introContentList.length; i++) {
+    if (introContentList[i]) {
+      const parent = introContentList[i].parentNode;
+      if (parent === null) {
+        break;
+      }
+      parent.removeChild(introContentList[i]);
+    }
+  }
+};
+
+export const generateExcerpt = (str: string) => {
+  return str.replace(/\n/g, ' ').replace(/\s\s/g, ' ').trim().substring(0, 152) + '...';
+};
+
+// tslint:disable-next-line: max-line-length
+const getPageType = (node: HTMLElement, values: DescriptionTemplateValues): PageTypes => {
+  const {parentType, chapterTitle, pageTitle} = values;
+  const nodeClasses = node.classList;
+  const nodeType = node.getAttribute('data-type');
+
+  if (nodeType === 'page') {
+    return 'page';
+  } else if (nodeClasses.contains('appendix')) {
+    return 'appendix';
+  } else if (
+    nodeClasses.contains('os-solution-container')
+      || nodeClasses.contains('os-solutions-container')
+      ) {
+    return 'answer-key';
+  } else if (parentType !== 'chapter' && parentType !== 'book') {
+    return 'eoc-sub-page';
+  } else if (pageTitle !== chapterTitle) {
+    return 'eoc-page';
+  } else {
+    return 'eob-page';
+  }
+};
+
+const getPageDescriptionFromContent = (node: HTMLElement): string | null => {
+  if (!node) {
+    return null;
+  }
+  removeIntroContent(node);
+  const firstSection = node.querySelector('section');
+  let para: Element | null = firstSection
+    ? firstSection.querySelector('p')
+    : node.querySelector('p');
+
+  if (!para || !para.textContent) {
+    return null;
+  }
+
+  // Find first <p> of 90+ chars.
+  if (para.textContent.length >= 90) {
+    const mathless = hideMath(para);
+    return mathless ? generateExcerpt(mathless) : null;
+  } else {
+    const text = para.textContent;
+    const next = para.nextElementSibling;
+    while (text && text.length < 90 && next) {
+      if (next.matches('p')) {
+        para = next;
+      }
+    }
```

Comment:
i think you could replace most of this logic by initially doing `querySelectorAll` up on line 115, and then doing `find` with your length criteria